### ISSUE TITLE
Saving and then reloading a model, optimizer, and data loader prepared with accelerate causes weird loss spikes

### ISSUE THREAD

**Issue by AntreasAntoniou** [huggingface/accelerate#1479]
### System Info

```Shell
- `Accelerate` version: 0.19.0
- Platform: Linux-5.4.0-148-generic-x86_64-with-glibc2.35
- Python version: 3.10.10
- Numpy version: 1.24.3
- PyTorch version (GPU?): 2.0.0 (True)
- System RAM: 448.57 GB
- GPU type: NVIDIA A100-SXM4-80GB
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: bf16
        - use_cpu: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I am using my own codebase, where I use accelerate as a means to enable easy bf16, gradient accumulation, gradient clipping, and in the near future FSDP. 

The relevant class that uses accelerate can be found here https://github.com/AntreasAntoniou/TALI/blob/d0553b80fecfa6fcad2520dd1f074c05689f23e2/tali/boilerplate.py#L431

The main issue can be seen in the attached figure. The spikes you see are whenever a node dies and the experiment resumes on a new node. I am suspicious that it's something relating to the optimizer reloading but I am not well versed with accelerate to debug the behaviour. Any help would be very appreciated. 

<img width="1120" alt="image" src="https://github.com/huggingface/accelerate/assets/10792502/fce86d10-cf07-4bc9-9e12-49cab600820e">

As you can see this is causing lots of performance loss that the model slowly recovers, but I would ideally like to not have to spend GPU time on that. 

### Expected behavior

The states should be saved and reloading and training should continue normally without odd optimization spikes.

**Comment by AntreasAntoniou** [comment]
I've resolved this I believe.
